### PR TITLE
Store Promotions: Add variation sales to list.

### PR DESCRIPTION
**Pull request by coderkevin**
This adds code to fetch variations for products and add on-sale
variations to the promotions list. Note that editing variation sales
will be covered by a PR after this one (see #20090)

This also begins using `getAllProducts` for some of the create/update
pages to prepare for de-duplicating promotion/product/coupon state.

To Test:
1. From wp-admin, add a sale price to a variation of a variable product.
2. View list of promotions, that variation sale should be visible in the list.

### CODE REVIEW

File: client/extensions/woocommerce/state/sites/promotions/helpers.js [hunk 1]
```diff
@@ -8,19 +8,32 @@ export function createPromotionFromProduct( product ) {
...
 		endDate,
 		productId,
+		parentId,
 	};
 }
 
+function createProductName( product ) {
+	if ( product.productId ) {
+		// It's a variation, append the options.
+		const options = product.attributes.map( attr => attr.option ).join( ', ' );
```
Comment (ryelle) [inline client/extensions/woocommerce/state/sites/promotions/helpers.js]: You could use [`formattedVariationName `](https://github.com/Automattic/wp-calypso/blob/master/client/extensions/woocommerce/lib/formatted-variation-name/index.js#L10) here, basically does the same thing but then we'd be consistent across Store.

Comment (coderkevin) [reply client/extensions/woocommerce/state/sites/promotions/helpers.js]: Oh, thanks!

**Comment by coderkevin** [comment]
Rebased onto current master.

**Comment by coderkevin** [comment]
Rebased on current master to resolve a couple conflicts.

**Comment by coderkevin** [comment]
Merging 2 other PRs into this one before merging to master. #20090 and #20377

**Pull request merged**